### PR TITLE
feature: add `flashbots_*` rpc

### LINKED ISSUE #27: Add flashbots_ rpc namespace support
ref https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#flashbots_getuserstatsv2

## TODO
* add types + rpc trait, see existing eth/mev.rs

### PR DESCRIPTION
**Pull request by optimiz-r**
depends on #33 because of https://github.com/paradigmxyz/mev-share-rs/pull/34/files#diff-3c0917e144af13736e5dd8c32c0ad1497fa59c20361b8454f0064e1a368871aaR102

closes #27

### CODE REVIEW

File: crates/mev-share-rpc-api/src/types.rs [hunk 1]
```diff
@@ -418,28 +418,97 @@ pub struct CancelPrivateTransactionRequest {
     pub tx_hash: H256,
 }
 
+// TODO(@optimiz-r): Revisit after <https://github.com/flashbots/flashbots-docs/issues/424> is closed.
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/types.rs:421]: lets track this as issue?

File: crates/mev-share-rpc-api/src/types.rs [hunk 2]
```diff
@@ -418,28 +418,97 @@ pub struct CancelPrivateTransactionRequest {
...
 /// Response for `flashbots_getBundleStatsV2` represents stats for a single bundle
 ///
 /// Note: this is V2: <https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#flashbots_getbundlestatsv2>
 ///
 /// Timestamp format: "2022-10-06T21:36:06.322Z"
+#[derive(Default, Debug, Clone, PartialEq)]
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/types.rs:427]:
```suggestion
#[derive(Default, Debug, Clone, PartialEq, Eq)]
```

File: crates/mev-share-rpc-api/src/types.rs [hunk 3]
```diff
@@ -418,28 +418,97 @@ pub struct CancelPrivateTransactionRequest {
...
+#[serde(rename_all = "camelCase")]
+pub struct StatsSeen {
+    /// boolean representing if this searcher has a high enough reputation to be in the high
+    /// priority queue
+    is_high_priority: bool,
+    /// representing whether the bundle gets simulated. All other fields will be omitted except
+    /// simulated field if API didn't receive bundle
+    is_simulated: bool,
+    /// time at which the bundle API received the bundle
+    received_at: String,
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/types.rs:484]: these should be public

File: crates/mev-share-rpc-api/src/types.rs [hunk 4]
```diff
@@ -418,28 +418,97 @@ pub struct CancelPrivateTransactionRequest {
...
+impl<'de> Deserialize<'de> for BundleStats {
+    fn deserialize<D>(deserializer: D) -> Result<BundleStats, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let map = serde_json::Map::deserialize(deserializer)?;
+
+        if map.get("receivedAt").is_none() {
+            Ok(BundleStats::Unknown)
+        } else if map["isSimulated"] == "false" {
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/types.rs:457]: this doesn't look like serialize and deserialize is symmetrical.
looking at `isSimulated`

File: crates/mev-share-rpc-api/src/types.rs [hunk 5]
```diff
@@ -418,28 +418,97 @@ pub struct CancelPrivateTransactionRequest {
...
-    pub considered_by_builders_at: Vec<ConsideredByBuildersAt>,
+    received_at: String,
+    /// indicates time at which each builder selected the bundle to be included in the target
+    /// block
+    #[serde(default = "Vec::new")]
+    considered_by_builders_at: Vec<ConsideredByBuildersAt>,
     /// indicates time at which each builder sealed a block containing the bundle
-    pub sealed_by_builders_at: Vec<SealedByBuildersAt>,
+    #[serde(default = "Vec::new")]
+    sealed_by_builders_at: Vec<SealedByBuildersAt>,
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/types.rs:511]: all of them should be pub

File: crates/mev-share-rpc-api/src/types.rs [hunk 6]
```diff
@@ -456,6 +525,38 @@ pub struct SealedByBuildersAt {
...
+    /// queue.
+    is_high_priority: bool,
+    /// The total amount paid to validators over all time.
+    #[serde(with = "u256_numeric_string")]
+    all_time_validator_payments: U256,
+    /// The total amount of gas simulated across all bundles submitted to Flashbots.
+    /// This is the actual gas used in simulations, not gas limit.
+    #[serde(with = "u256_numeric_string")]
+    all_time_gas_simulated: U256,
+    /// The total amount paid to validators the last 7 days.
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/types.rs:544]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
this looks great already

**Comment by optimiz-r** [comment]
spot on, taking care of these immediately

File: crates/mev-share-rpc-api/src/types.rs (hunk 1 above)
Comment (optimiz-r) [reply crates/mev-share-rpc-api/src/types.rs:421]: as in open new issue in repo linking to this and get rid of TODO line?

**Review by mattsse — APPROVED** [review]
that's awesome, tysm!

**Pull request merged**